File: docs/dp_tabulation/1359_hard/pickup_deliveries.cpp
#include <cmath>
using namespace std;

int countValidSequences(int totalOrders) // LeetCode Q.1359.
{
    long long modulo = pow(10, 9) + 7; // Long long prevents overflow.

    // Base case when n is 1: pickup 1 followed by delivery 1.
    long long validSequences = 1;

    for (int prevOrderNum = 1; prevOrderNum <= totalOrders - 1; prevOrderNum++) {
        int spotsCount = 2 * prevOrderNum + 1;

        // Sum from 1 to spots count.
        int multiple = spotsCount * (spotsCount + 1) / 2;

        validSequences *= multiple;
        validSequences %= modulo;
    }

    return validSequences;
}

File: docs/dp_tabulation/1359_hard/pickup_deliveries.py
def count_valid_sequences(total_orders: int) -> int:
    modulo = 10**9 + 7

    # Base case when n is 1: pickup 1 followed by delivery 1.
    valid_sequences = 1

    for prev_order_num in range(1, total_orders):
        spots_count = 2 * prev_order_num + 1

        # Sum from 1 to spots count.
        multiple = spots_count * (spots_count + 1) // 2

        valid_sequences *= multiple
        valid_sequences %= modulo

    return valid_sequences
